packages/json-parser-products/src/mappings.js

```diff
@@ -0,0 +1,172 @@
+
+  constructor ({
+    fillAllRows = false,
+    categoryBy = 'name',
+    lang = 'en',
+    multiValueDelimiter = ';',
+  }: Object = {}) {
+    this.fillAllRows = fillAllRows
+    this.categoryBy = categoryBy
+    this.lang = lang
+    this.multiValDel = multiValueDelimiter
+  }
+
+  run (product: ResolvedProdProj) {
+    const mergedVarProducts = ProductMapping._mergeVariants(product)
+    const variantsWithProductInfo = ProductMapping._spreadProdOnVariants(
+      mergedVarProducts, this.fillAllRows)
+    const a = variantsWithProductInfo.map(
+      (variant: SingleVariantPerProduct) => (
+        this._mapProperties(variant)
+      ))
+    const flatProducts = a.map(flatten)
+    return flatProducts
+  }
+
+  static _mergeVariants (product: ResolvedProdProj): ProdWithMergedVariants {
+    const variant = [product.masterVariant, ...product.variants]
+    const mergedVariants = { ...product, variant }
+    delete mergedVariants.masterVariant
+    delete mergedVariants.variants
+    return mergedVariants
+  }
+
+  static _spreadProdOnVariants (
+    product: ProdWithMergedVariants,
+    fillAllRows: boolean,
+  ): Array<SingleVariantPerProduct> {
+    if (fillAllRows)
+      return product.variant.map((eachVariant: Variant): Object => (
+        { ...product, variant: eachVariant }
+      ))
+
+    const productWithVariants: Array<Object> = product.variant.map((
+      eachVariant: Variant,
+    ): SingleVariantPerProduct => (
+      { variant: eachVariant }
+    ))
+    productWithVariants[0] = { ...product, ...productWithVariants[0] }
+    return productWithVariants
+  }
+
+  _mapProperties (product: SingleVariantPerProduct): MappedProduct {
+    return reduce(product, (
+      acc: Object,
+      value: any,
+      property: string,
+    ): Object => {
+      switch (property) {
+        case 'productType': {
+          acc[property] = value.name
+          break
+        }
+        case 'state': {
+          acc[property] = value.key
+          break
+        }
+        case 'taxCategory': {
+          acc[property] = value.key || value.name
+          break
+        }
+        case 'categories': {
+          acc[property] = ProductMapping._mapCategories(
+            value, this.categoryBy, this.multiValDel, this.lang)
+          break
+        }
+        case 'categoryOrderHints': {
+          if (!isEmpty(value))
+            acc[property] = Object.keys(value).map((key: string): string => (
+              `${key}:${value[key]}`
+            )).join(this.multiValDel)
+          break
+        }
+        case 'searchKeywords': {
+          // TODO: Handle the searchKeywords
+          break
+        }
+        case 'version': {
+          // Remove version number
+          break
+        }
+        case 'variant': {
+          if (!isEmpty(value.attributes)) {
+            acc.attr = {}
+
+            value.attributes.forEach((attribute: Object) => {
+              acc.attr[attribute.name] = attribute.value.key || attribute.value
```

Comment:
All attributes have a `value` property regardless of the type. If the `value` property is an object, then it always has a unique `key`

Reply:
@wizzy25 I checked few projects and I found, for example, this `number` attribute:
```json
{
    "name": "attr-number",
   "value": 8300138
}
```
Which would cause an error when retrieving `attribute.value.key` (same as boolean or a simple text would do)
Then there is, for example, this `ltext` object without the key:
```json
{
  "name": "attr-ltext",
  "value": {
    "de": "PCE"
  }
}
```

Reply:
@junajan If a field has no `key` value, like the examples you sighted, undefined will be returned, and the attribute will be set to `value` instead of `value.key` (It will not throw)
`attribute.value.key || attribute.value`
In the latter case, on flattening we will simply get a localised field in our csv after flattening, which is what we want.
You can see one of the tests for it [here](https://github.com/commercetools/nodejs/pull/326/files/4ea77c1801075e182193f1f9cf0171eea052a1a2#diff-f675f797a8217a80c1646e0340b26671R168) and [here](https://github.com/commercetools/nodejs/pull/326/files/4ea77c1801075e182193f1f9cf0171eea052a1a2#diff-f675f797a8217a80c1646e0340b26671R210)